### PR TITLE
Integrate PurgeCSS directly into Tailwind

### PR DESCRIPTION
**Pull request by adamwathan**
This PR adds a new `purge` config option for purging unused CSS directly from within Tailwind instead of having to pull in and configure PurgeCSS manually.

The API is much more opinionated and slimmed down by default, optimizing for the 95% use case:

```js
// tailwind.config.js
module.exports = {
  purge: [
    './src/**/*.html',
    './src/**/*.html',
  ],
  theme: {},
  variants: {},
  plugins: [],
}
```

The `purge` option accepts an array of filenames/globs to scan for classes, and that's it. This should be enough for most people, but if you are one of those people who just can't help themselves and has to fiddle with everything, you can fine-tune the configuration however you like (keep reading for more on that).

Things to note:

- **This option is only enabled when `NODE_ENV` is `production`.** If you want to enable it explicitly based on some other conditional logic, you can use an object syntax and the `enabled` property, and specify your templates using the `content` property:
  
    ```js
    // tailwind.config.js
    module.exports = {
      purge: {
        enabled: true,
        content: ['./src/**/*.html'],
      },
      // ...
    }
    ```

- **Only components and utilities are purged by default.** Our base styles are not purged, and no custom CSS authored in your CSS files is purged, including any imports to outside libraries.

    This is to make sure you don't accidentally run into common pitfalls, like forgetting to scan your root HTML template (which is difficult with frameworks like Next.js since it is buried in `node_modules`) or accidentally purging styles that come from whatever datepicker library you've pulled in as an npm dependency.

    This means we are not purging _absolutely everything_ that could be safely purged, but you get 99% of the benefit with much less manual tuning. You will basically never need to whitelist anything using this approach.

    Essentially we make your CSS file look like this:

    ```css
    /* purgecss start ignore */
    @tailwind base;

    /* purgecss end ignore */
    @tailwind components;
    /* purgecss start ignore */

    .btn {
      @apply bg-blue-500 px-4 py-2 text-white;
    }

    .alert {
      @apply border p-4 rounded;
    }

    /* purgecss end ignore */
    @tailwind utilities;
    /* purgecss start ignore */

    .custom-utility {
      foo: bar;
    }
    /* purgecss end ignore */
    ```

    **If you really want to purge everything, you can do so using `mode: 'all'`**:

    ```js
    // tailwind.config.js
    module.exports = {
      purge: {
        mode: 'all',
        content: ['./src/**/*.html'],
      },
      // ...
    }
    ```

    You're much more likely to accidentally remove important styles if you do this but I'm not the cops — if it tickles your pickle go for it.

- **We are using a new extractor strategy that tries to be as permissive as possible.**

    Here is the new extractor we're using:
    
    ```js
    defaultExtractor: content => {
      // Capture as liberally as possible, including things like `h-(screen-1.5)`
      const broadMatches = content.match(/[^<>"'`\s]*[^<>"'`\s:]/g) || []

      // Capture classes within other delimiters like .block(class="w-1/2") in Pug
      const innerMatches = content.match(/[^<>"'`\s.()]*[^<>"'`\s.():]/g) || []

      return broadMatches.concat(innerMatches)
    }
    ```

    ...and here is the sample HTML we test our purging against:

    ```html
    <!-- Basic HTML -->
    <div class="bg-red-500 md:bg-blue-300 w-1/2"></div>

    <!-- Vue dynamic classes -->
    <span :class="{ block: enabled, 'md:flow-root': !enabled }"></span>

    <!-- JSX with template strings -->
    <script>
      function Component() {
        return <div class={`h-screen`}></div>
      }
    </script>

    <!-- Custom classes with really weird characters -->
    <div class="min-h-(screen-4) bg-black! font-%#$@ w-(1/2+8)"></div>

    <!-- Pug -->
    span.inline-grid.grid-cols-3(class="px-1.5")
      .col-span-2
        Hello
      .col-span-1.text-center
        World!
    ```

    It works really well and I would be pretty surprised if you ever needed to customize this.

If you want more control than what we offer, you can pass any options you want directly to PurgeCSS using the `options` property:

```js
// tailwind.config.js
module.exports = {
  purge: {
    enabled: true,
    content: ['./src/**/*.html'],

    // These options are passed through directly to PurgeCSS
    options: {
      whitelist: ['bg-red-500', 'px-4'],
    }
  },
  // ...
}
```

Any options passed to PurgeCSS will override our own options, so for example if you pass `content` it will override our `content` property at the top-level.

Happy purging 🤮

### CODE REVIEW

**Comment by MrJmpl3** [comment]
Please, don't forget add a example with prefix 😄

**Comment by utkarshkukreti** [comment]
> Right now this won't work with Pug templates because we are allowing . characters in class names (Tailwind UI has classes like ml-0.5 for example). This is unfortunate but I can't think of a solution that allows both :/

How about matching the regex you currently have AND the same thing without dots?

```
> const c = 'foo bar.baz';
undefined
> c.match(/[^<>"'`\s]*[^<>"'`\s:]/g).concat(c.match(/[^<>"'`\s]*[^<>"'`\s:]/g))
[ 'foo', 'bar.baz', 'foo', 'bar.baz' ]
> c.match(/[^<>"'`\s]*[^<>"'`\s:]/g).concat(c.match(/[^<>"'`\s.]*[^<>"'`\s:.]/g))
[ 'foo', 'bar.baz', 'foo', 'bar', 'baz' ]
```

**Comment by h311x** [comment]
JSX uses “className” not “class” :)
Also would it be capable of allowing us to use conditional class names in vue templates like this:

:class="{[someComputedValueThatReturnsAString]: true}"

**Comment by adamwathan** [comment]
> JSX uses “className” not “class” :)

Thankfully not important for this.

> Also would it be capable of allowing us to use conditional class names in vue templates like this:
> :class="{[someComputedValueThatReturnsAString]: true}"

Only if the computed value contains the entire class string and isn't dynamically calculated with weird string interpolation or anything. All of the regular limitations of PurgeCSS are unchanged here, read our guide on writing purgeable HTML:

https://tailwindcss.com/docs/controlling-file-size/#writing-purgeable-html

**Comment by adamwathan** [comment]
> How about matching the regex you currently have AND the same thing without dots?

@utkarshkukreti Good idea! Updated it to support Pug a bit better using your solution 👍

**Comment by brandonpittman** [comment]
@adamwathan  Will this pass through whitelisting?

**Comment by kirkbushell** [comment]
haha this is great. I just added a little script on my forge deployments to do exactly this!

**Comment by klaasgeldof** [comment]
Great improvement!

Unfortunately, the `ignore` annotations don't work when importing external css into sass. For example, I have this in `app.scss` one of my projects:

```css
@tailwind base;

@tailwind components;

/* purgecss start ignore */
@import "~toastr/toastr.scss";
@import "~@fancyapps/fancybox/dist/jquery.fancybox.css";
/* purgecss end ignore */

@tailwind utilities;
```

But this still strips all of the css of toastr and fancybox. If you have a solution for this, that would be great 😃

Otherwise I guess it's still possible to pass `whitelistPatterns`?

**Comment by dominikzogg** [comment]
Using ReactJS would mean add *.jsx or *.tsx to the list?

Using VueJs would mean add *.Vue to the list?

No other changes required?

**Comment by tbillington** [comment]
> no custom CSS authored in your CSS files is purged

Could this possibly be configurable?

**Comment by OwenMelbz** [comment]
This is awesome 🤩

What’s the chances of you allowing to pass in a whitelist to ignore? I feel like many people use the whitelist... especially on CMS driven websites where classes might come from the database 😇

**Comment by diwms** [comment]
Anyone already trying this with Gatsby? Please, share your feedback if yes!

**Comment by berndartmueller** [comment]
Very nice feature! Thx! 

I did a quick test with your regex pattern and pug syntax and it turns out that following pug code does not properly support purging css class names:

```
span.inline-grid.grid-cols-3(class="px-1.5")
  .col-span-2
    Hello
  .col-span-1.text-center
    World! 
```

Output of regex matching is following:
`["span.inline-grid.grid-cols-3(class=", "px-1.5", ")", ".col-span-2", "Hello", ".col-span-1.text-center", "World!", "span", "inline-grid", "grid-cols-3(class=", "px-1", "5", ")", "col-span-2", "Hello", "col-span-1", "text-center", "World!"]`

As you can see, `grid-cols-3` is wrongly extracted.

I had this issue already a few weeks ago and I came up with the solution to do the class extraction in multiple passes (following is postcss-purgecss config):

```
defaultExtractor: content => {
    const firstPassMatches = content.match(/[\w-/:]+(?<!:)/g) || [];
    const secondPassMatches = content.match(/[\w-/.:]+(?<!:)/g) || [];
    const combined = firstPassMatches.concat(secondPassMatches);
    const unique = Array.from(new Set(combined));

    return unique;
  },
```

**Comment by adamwathan** [comment]
> But this still strips all of the css of toastr and fancybox. If you have a solution for this, that would be great

@klaasgeldof The solution in your case is to use the preservable comment syntax for Sass:

```css
/** purgecss start ignore */
```

Note the double asterisk at the beginning 👍 

Either way this issue won't impact this feature, these comments are added dynamically by Tailwind during the PostCSS phase, which is well after Sass is done looking at the files.

**Comment by adamwathan** [comment]
> Using ReactJS would mean add *.jsx or *.tsx to the list?
> 
> Using VueJs would mean add *.vue to the list?
> 
> No other changes required?

@dominikzogg Correct, you just pass an array of paths to where all of your templates are and you're done (as long as you follow the rules of [writing purgeable HTML](https://tailwindcss.com/docs/controlling-file-size/#writing-purgeable-html)).

**Comment by dominikzogg** [comment]
> @dominikzogg Correct, you just pass an array of paths to where all of your templates are and you're done (as long as you follow the rules of [writing purgeable HTML](https://tailwindcss.com/docs/controlling-file-size/#writing-purgeable-html)).

@adamwathan perfect

**Comment by adamwathan** [comment]
I've pushed an update that makes things more configurable. Details below!

### Passing options directly to PurgeCSS

Now you can pass an `options` object which will be passed directly to PurgeCSS so you have access to any other PurgeCSS features you want like whitelisting, overriding the extractor, whatever you want.

```js
// tailwind.config.js
module.exports = {
  purge: {
    content: ['./src/**/*.html'],

    // These options are passed directly to PurgeCSS
    options: {
      whitelist: ['bg-red-500', 'px-4']
    }
  },
  // ...
}
```

I've also renamed the `paths` option to `content` to match PurgeCSS.

### Better Pug support

I've improved the matching strategy a bit more to better handle Pug. Works pretty great now, even on stuff like this:

```pug
span.inline-grid.grid-cols-3(class="px-1.5")
  .col-span-2
    Hello
  .col-span-1.text-center
    World!
```

### New `mode` feature

As I mentioned in the original post, this PurgeCSS integration is designed to be fairly conservative by default and only purges styles generated by `@tailwind components` and `@tailwind utilities` to make it easier to integrate with third-party tools and to avoid accidentally stripping base styles when working with frameworks that hide their base HTML template in `node_modules` like Next.js.

If you want to purge _everything_ (at your own risk of course), you can now do so by setting `mode` to `all`:

```js
// tailwind.config.js
module.exports = {
  purge: {
    mode: 'all'
    content: ['./src/**/*.html'],
  },
  // ...
}
```

The default mode is "conservative" in case you're ever inclined to set it explicitly:

```js
// tailwind.config.js
module.exports = {
  purge: {
    mode: 'conservative'
    content: ['./src/**/*.html'],
  },
  // ...
}
```

**Comment by hkhanna** [comment]
This is huge. For folks (me) who accidentally purged things like the nprogress css styles and then are dumbfounded why the progress bar was not showing up, this will ease hours of frustration.

**Comment by Haqverdi** [comment]
is it will be `enabled` by default?

**Comment by adamwathan** [comment]
> is it will be enabled by default?

@Haqverdi Only if you have actually provided a list of paths and `NODE_ENV` is `production`. If you leave the `purge` option out of your config entirely it will never be enabled.

**Comment by tpmarc** [comment]
This is awesome. Can't wait to use it!

**Comment by dominikzogg** [comment]
@adamwathan updated to 1.4 and adapted the config, works like a charm (react, vue)

https://github.com/chubbyphp/petstore-react/commit/5c27bcb0c379a11eda73f8eed1546b8565182e14
https://github.com/chubbyphp/petstore-vue/commit/92fd622aea716cb3fb283a2582c675f208455465

**Comment by Waleedov** [comment]
I use gulp, how is this going to work if I don't have production command?

**Comment by adamwathan** [comment]
@Waleedov Set the NODE_ENV variable to `production` when running your gulp script.

```
NODE_ENV=production gulp
```

**Comment by nadnoslen** [comment]
Thanks @adamwathan, this is awesome.

**Comment by nhulongctk35** [comment]
Thanks @adamwathan.

**Comment by shirazz** [comment]
```
const purgecssConfig = {
  // The paths to all of the template files in the project
  content: ["./components/**/*.jsx", "./pages/**/*.jsx"],
  whitelistPatterns: [/^PhoneInput/, /^Toastify/, /^alice-carousel/],
  whitelistPatternsChildren: [/^Toastify/],

  // This is the function used to extract class names from your templates
  defaultExtractor: (content) => {
    // Capture as liberally as possible, including things like `h-(screen-1.5)`
    const broadMatches = content.match(/[^<>"'`\s]*[^<>"'`\s:]/g) || [];
    // Capture classes within other delimiters like .block(class="w-1/2") in Pug
    const innerMatches = content.match(/[^<>"'`\s.()]*[^<>"'`\s.():]/g) || [];
    return broadMatches.concat(innerMatches);
  },
};

module.exports = {
  plugins: {
    "postcss-import": {},
    tailwindcss: {},
    "postcss-nested": {},
    "postcss-custom-properties": {},
    autoprefixer: {},
    "@fullhuman/postcss-purgecss":
      process.env.NODE_ENV === "production" ? purgecssConfig : false,
  },
};
```

In my NextJS app, I have a `postcss.config.js` file like the above. Updated the tailwind to new version which has the purgeCSS support inbuilt. I tried removing the postcss-purgecss plugin and added the purge css configuration in my tailwind.config.js Unfortunately it's not working. Can anyone please guide me here to make this work.

**Comment by adamwathan** [comment]
Can you create a project that reproduces the issue so I can better help troubleshoot?

**Comment by shirazz** [comment]
I will check in details again and get back to you soon. Thanks @adamwathan for your quick response.

**Comment by shirazz** [comment]
@adamwathan Really sorry this was purely my fault. For some reason my tailwind was not updated to 1.4. Once i did it, everything worked as a charm. Thank you very much for adding this feature out of the box. 

BTW. the only issue which i face is, while using a custom variant,  [tailwind-dir](https://github.com/RonMelkhior/tailwindcss-dir) where my `ltr:` styles are purged. I haven't gone deep into this. but the issue is there. I can provide a repo if it helps.

**Comment by divdax** [comment]
Today i updated tailwind 1.4 to use purgecss. One thing i've noticed is that `.mr-1.5` was purged. The class is in use in one of my blade files.

**Comment by klaasgeldof** [comment]
@divdax Had the same thing, this fixed it for me:

```js
purge: {
  options: {
    whitelistPatterns: [/^.[^\.]*\.[^\. "']*/],
  },
},
```

@adamwathan Maybe include this pattern by default?

**Comment by divdax** [comment]
It works! Thank you @klaasgeldof! 🙏 And yes... this pattern should be included by default.

**Comment by devcircus** [comment]
Not sure the framework should include white list patterns out of the box. That should probably be handled via an extractor pattern, like the one suggested by tailwindui.

**Comment by adamwathan** [comment]
That class should not be purged using our default settings, we even have a test for that here:

https://github.com/tailwindcss/tailwindcss/blob/master/__tests__/purgeUnusedStyles.test.js#L49

Can you create a project that reproduces the issue?

**Comment by klaasgeldof** [comment]
I was wrong. Before Tailwind had purge baked in, the classes with dots were purged and I had to add the pattern. I copied my patterns (including this one) to the Tailwind purge config because I supposed it would behave the same. But now I removed the pattern and it works... @divdax Check if there is no other purging going on in your production build apart from the Tailwind purge.

**Comment by agcty** [comment]
@adamwathan What do you mean with:
> * Essentially we make your CSS file look like this:
>   ```css
>   /* purgecss start ignore */
>   @tailwind base;
>   
>   /* purgecss end ignore */
>   @tailwind components;
>   /* purgecss start ignore */
>   
>   .btn {
>     @apply bg-blue-500 px-4 py-2 text-white;
>   }
>   
>   .alert {
>     @apply border p-4 rounded;
>   }
>   
>   /* purgecss end ignore */
>   @tailwind utilities;
>   /* purgecss start ignore */
>   
>   .custom-utility {
>     foo: bar;
>   }
>   /* purgecss end ignore */
>   ```

We still need to add some sort of tailwind.scss file and import it globally so all our components can use tailwind classes right? Would we copy the contents of your comment? Or could we just use:

```
@tailwind base;
@tailwind components;
@tailwind utilities;
```
and the right stuff is purged automatically without having to use /* purgecss start ignore */ comments?

**Comment by adamwathan** [comment]
You don't have to copy the CSS I pasted there, it's just explaining how it's working conceptually to people who are used to configuring PurgeCSS with the ignore comments manually.

**Comment by claudiodekker** [comment]
Super cool to see this as part of the core! 

For those who want to combine this with Laravel (and possibly also for those migrating from `spatie-mix-purgecss`), here's the configuration set `spatie-mix-purgecss` was using, which might give you a good starting point:

```js
module.exports = {
  purge: {
      // enabled: true,
      content: [
          "app/**/*.php",
          "resources/**/*.html",
          "resources/**/*.js",
          "resources/**/*.jsx",
          "resources/**/*.ts",
          "resources/**/*.tsx",
          "resources/**/*.php",
          "resources/**/*.vue",
          "resources/**/*.twig",
      ],
      options: {
          // defaultExtractor: (content) => content.match(/[\w-/.:]+(?<!:)/g) || [],
          whitelistPatterns: [/-active$/, /-enter$/, /-leave-to$/, /show$/],
      }
  },
  theme: {
    extend: {},
  },
  variants: {},
  plugins: [],
}

```

If you want 100% the same exact behavior as `laravel-mix-purgecss`, you can uncomment the `defaultExtractor` option. If you're not sure and just want something that works, just stick with this default, as it's pretty much tailor-fit to TailwindCSS.

Hope this helps!

**Comment by cygnusross** [comment]
> Super cool to see this as part of the core!
> 
> For those who want to combine this with Laravel (and possibly also for those migrating from `spatie-mix-purgecss`), here's the configuration set `spatie-mix-purgecss` was using, which might give you a good starting point:
> 
> ```js
> module.exports = {
>   purge: {
>       // enabled: true,
>       content: [
>           "app/**/*.php",
>           "resources/**/*.html",
>           "resources/**/*.js",
>           "resources/**/*.jsx",
>           "resources/**/*.ts",
>           "resources/**/*.tsx",
>           "resources/**/*.php",
>           "resources/**/*.vue",
>           "resources/**/*.twig",
>       ],
>       options: {
>           // defaultExtractor: (content) => content.match(/[\w-/.:]+(?<!:)/g) || [],
>           whitelistPatterns: [/-active$/, /-enter$/, /-leave-to$/, /show$/],
>       }
>   },
>   theme: {
>     extend: {},
>   },
>   variants: {},
>   plugins: [],
> }
> ```
> 
> If you want 100% the same exact behavior as `laravel-mix-purgecss`, you can uncomment the `defaultExtractor` option. If you're not sure and just want something that works, just stick with this default, as it's pretty much tailor-fit to TailwindCSS.
> 
> Hope this helps!

Very useful, thank you!

**Comment by cll0ud** [comment]
I got it working fine on `nuxt.js` with Vuetify. I had to disable purgeCSS from `nuxt.config.js` because it was breaking Vuetify's style, after that it worked like a charm using the following config:

```js
// tailwind.config.js
module.exports = {
  prefix: 'tw-',
  purge: {
    content: [
      './components/**/*.vue',
      './layouts/**/*.vue',
      './pages/**/*.vue'
    ],
    options: {
      whitelist: ['antialiased', 'font-sans']
    }
  },
  important: true,
  theme: {},
  variants: {},
  plugins: []
}
```

I noticed a little problem: I added the prefix "tw-" and when checking the generated css, I noticed that it had included ".tw-container" classes from Tailwind but I haven't used those at all. Thing is, Vuetify's got their own ".container" so it got ignored by the extractor even though I was using the prefix. Not a major problem but still unexpected result, I guess?

BTW, thanks for this feature, I love it!

**Comment by MalvinJay** [comment]
This saved me two nights of shame. 😢 👍

**Comment by virgiltu** [comment]
I just changed to tailwindcss and that seems to require laravel mix purgecss.

My website is a website builder, once the website is build javascript will take the HTML and post it in a variable via request.

At the controller level I take that variable and save it to the database in a filed called PageHtml.

What I cant figure out, How can I take the variable that contains the page HTML and run it through laravel mix exporting the css to a variable or file. Ideal will be to return the css to a variable so I can also save it in the database with the page.

In short anybody knows of a way to run PurgeCss for Tailwind in Laravel but on a variable that contains HTML and output the css back into a variable?

I really appreciate the help.

**Comment by virgiltu** [comment]
This works great in a single app system. However in a CMS where you have lets say a User Front end and Admin section, how can you split the two css files. 

An an example 

`purge: {
    enabled: true,
    content: [
     './admin/**/Resources/views/*.blade.php',
     './users/**/Resources/views/*.blade.php'
       ],
  },`

Is there a way to get the css for each page defined?

**Comment by OwenMelbz** [comment]
Do you have 2 css files output from the same tailwind config?

Also, it's probably not recommended to use

purge.enabled = true

You should try use NODE_ENV as production and it's automatic then

**Comment by virgiltu** [comment]
> Do you have 2 css files output from the same tailwind config?
> 
> Also, it's probably not recommended to use
> 
> purge.enabled = true
> 
> You should try use NODE_ENV as production and it's automatic then

Thank you for the response, I think I figured out what I was looking to do to a point 

In order to create a mix file for backend and frontend each (and other entry points if needed),I need to  adjust package.json,  however this does not really fix my tailwind problem, maybe I need to create multiple tailwind config files. 

```
"scripts": {
    "dev": "npm run development",
    "development": "cross-env process.env.section=website NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
    "watch": "cross-env process.env.section=website NODE_ENV=development node_modules/webpack/bin/webpack.js --watch --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
    "prod": "npm run production",
    "production": "cross-env process.env.section=website NODE_ENV=production node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
    "admin-dev": "npm run admin-development",
    "admin-development": "cross-env process.env.section=admin NODE_ENV=development node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
    "admin-watch": "cross-env process.env.section=admin NODE_ENV=development node_modules/webpack/bin/webpack.js --watch --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js",
    "admin-prod": "npm run admin-production",
    "admin-production": "cross-env process.env.section=admin NODE_ENV=production node_modules/webpack/bin/webpack.js --progress --hide-modules --config=node_modules/laravel-mix/setup/webpack.config.js"
  },
```
webpack.mix.js:

```
let mix = require('laravel-mix');

if (process.env.section) {
  require(`${__dirname}/webpack.mix.${process.env.section}.js`);
}
```
create webpack.mix.website.js:

```
let { mix } = require('laravel-mix');

mix
  .setPublicPath(path.normalize('public_html/assets/website'))
  .less('resources/assets/website/less/website.less', 'css/style.css')
  .options({
    processCssUrls: false
  })
  .js('resources/assets/website/js/website.js', 'js/global.js')
;
```
create webpack.mix.admin.js:

```
let mix = require('laravel-mix');

mix
  .setPublicPath(path.normalize('public_html/assets/admin'))
  .options({
    processCssUrls: false
  })
  .js('resources/assets/admin/js/admin.js', 'js/global.js')
  .less('resources/assets/admin/less/admin.less', 'css/style.css')
;
```

**Comment by virgiltu** [comment]
I have to point out, the above solution still does not fix the fundamental problem. As many know google will rate a website based on many aspects and one is Unused Css. So even if you run a purgecss and make the file smaller it will still have classes from other pages so you will still be penalized by google. In reality the solution would be to generate one css file per page each page on a website.      

However it is still better than other frameworks.

**Comment by mklueh** [comment]
@virgiltu do you have a source for your statement?

**Comment by virgiltu** [comment]
> @virgiltu do you have a source for your statement?

Just use the google insight tool or lighthouse in dev tools.  You will see google taking points for having unused CSS on the page.
